sql/handler.cc

```diff
@@ -2067,6 +2093,18 @@ static my_bool xarecover_handlerton(THD *unused, plugin_ref plugin,
     {
       sql_print_information("Found %d prepared transaction(s) in %s",
                             got, hton_name(hton)->str);
+#ifdef WITH_WSREP
+      /* If wsrep_on=ON, XIDs are first ordered and then the range of
+         recovered XIDs is checked for continuity. All the XIDs which
+         are in continuous range can be safely committed if binlog
+         is off since they have already ordered and certified in the
+         cluster. */
+      my_xid wsrep_limit= 0;
+      if (WSREP_ON)
+      {
+        wsrep_limit= wsrep_order_and_check_continuity(info->list, got);
```

Comment:
Regardless of binlog settings the prepare step may now be executed out of order for transactions because the GTID is assigned in `wsrep_before_prepare()` but the commit ordering happens in `wsrep_before_commit()`. Therefore the XIDs may be returned from storage engine out of order.